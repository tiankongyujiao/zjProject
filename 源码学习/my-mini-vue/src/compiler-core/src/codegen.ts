import { NodeTypes } from "../../parse-core/src/ast";
import {
  TO_DISPLAY_STRING,
  helperMapName,
  CREATE_ELEMENT_BLOCK,
} from "./runtimeHelpers";

export function generate(ast) {
  const context = createCodegenContext();

  const { push } = context;

  genFunctionPreamble(ast, context);

  const functionName = "render";
  const args = ["_ctx", "_cache"];
  const signature = args.join(", ");

  push(`function ${functionName}(${signature}) {`);
  push(`return `);

  genNode(ast.codegenNode, context);

  push("}");

  return {
    code: context.code,
  };
}

function genFunctionPreamble(ast, context) {
  const { push, helper } = context;
  const VueBinging = "Vue";
  const aliasHelper = (s) => `${helper(s)}: _${helper(s)}`;
  if (ast.helpers.length > 0) {
    push(`const { ${ast.helpers.map(aliasHelper).join(",")} } = ${VueBinging}`);
    push("\n");
  }
  push("return ");
}

function genNode(node: any, context) {
  switch (node.type) {
    case NodeTypes.TEXT:
      genText(node, context);
      break;
    case NodeTypes.INTERPOLATION:
      genInterpolation(node, context);
      break;
    case NodeTypes.SIMPLE_EXPRESSION:
      genExpression(node, context);
      break;
    case NodeTypes.ELEMENT:
      genElement(node, context);
      break;
    default:
      break;
  }
}

function genElement(node, context) {
  const tag = node.tag;
  const { push, helper } = context;
  push(`${helper(CREATE_ELEMENT_BLOCK)}("${tag}")`);
}

function genExpression(node: any, context: any) {
  const { push } = context;
  push(`${node.content}`);
}

function genInterpolation(node, context) {
  const { push, helper } = context;
  push(`${helper(TO_DISPLAY_STRING)}(`);
  genNode(node.content, context);
  push(")");
}

function genText(node, context) {
  const { push } = context;
  push(`'${node.content}'`);
}

function createCodegenContext() {
  const context = {
    code: "",
    push(source) {
      context.code += source;
    },
    helper: (key) => {
      return `_${helperMapName[key]}`;
    },
  };

  return context;
}
