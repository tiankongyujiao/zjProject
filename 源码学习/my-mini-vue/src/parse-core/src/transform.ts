export function transform(root: any, options) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);
}
// 生成上下文
function createTransformContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
  };
  return context;
}

function traverseNode(node, context) {
  const nodeTransforms = context.nodeTransforms;
  for (let i = 0; i < nodeTransforms.length; i++) {
    nodeTransforms[i](node);
  }
  traverseChildren(node, context);
}

function traverseChildren(node, context) {
  const children = node.children;
  if (children) {
    for (let i = 0; i < children.length; i++) {
      const n = children[i];
      traverseNode(n, context);
    }
  }
}
