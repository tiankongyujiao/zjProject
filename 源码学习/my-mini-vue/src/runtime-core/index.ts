export { h } from "./h";
export { renderSlots } from "./helper/renderSlots";
export { createTextVNode } from "./vnode";
export { getCurrentInstance } from "./component";
export { provide, inject } from "./apiInject";
export { createApp } from '../runtime-dom/index'
export {createRenderer} from './render'
