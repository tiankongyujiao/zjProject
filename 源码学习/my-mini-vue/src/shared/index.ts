export * from "./toDisplayString";

export const extend = Object.assign;

export const isObject = (value) => {
  return value !== null && typeof value === "object";
};

export const isString = (value) => {
  return typeof value === "string";
};

export const hasChanged = (val, newVal) => {
  return !Object.is(val, newVal);
};

export const hasOwn = (val, key) =>
  Object.prototype.hasOwnProperty.call(val, key);

const capitalize = (event) => event.charAt(0).toUpperCase() + event.slice(1);

export const camelize = (str: string) => {
  return str.replace(/-(\w)/g, (_, c: string) => {
    return c ? c.toUpperCase() : "";
  });
};

export const toHandlerKey = (str) => (str ? "on" + capitalize(str) : "");
