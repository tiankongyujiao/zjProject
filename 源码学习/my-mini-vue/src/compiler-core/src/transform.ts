import { NodeTypes } from "../../parse-core/src/ast";
import { TO_DISPLAY_STRING } from "./runtimeHelpers";

export function transform(root: any, options = {}) {
  const context = createTransformContext(root, options);
  traverseNode(root, context);
  createRootCodegen(root);

  root.helpers = [...context.helpers.keys()];
}

function createRootCodegen(root) {
  const child = root.children[0];
  if (child.type === NodeTypes.ELEMENT) {
    root.codegenNode = child.codegenNode;
  } else {
    root.codegenNode = root.children[0];
  }
}
// 生成上下文
function createTransformContext(root, options) {
  const context = {
    root,
    nodeTransforms: options.nodeTransforms || [],
    helpers: new Map(),
    helper: (key) => {
      context.helpers.set(key, 1);
    },
  };
  return context;
}
// 转换节点
function traverseNode(node, context) {
  const nodeTransforms = context.nodeTransforms;
  const exitFns: any = [];
  for (let i = 0; i < nodeTransforms.length; i++) {
    const transform = nodeTransforms[i];
    const onExit = transform(node, context);
    if (onExit) exitFns.push(onExit);
  }
  switch (node.type) {
    case NodeTypes.INTERPOLATION:
      context.helper(TO_DISPLAY_STRING);
      break;

    case NodeTypes.ROOT:
    case NodeTypes.ELEMENT:
      traverseChildren(node, context);
      break;
    default:
      break;
  }

  let i = exitFns.length;
  while (i--) {
    exitFns[i]();
  }
}
// 处理children
function traverseChildren(node, context) {
  const children = node.children;
  if (children) {
    for (let i = 0; i < children.length; i++) {
      const n = children[i];
      traverseNode(n, context);
    }
  }
}
