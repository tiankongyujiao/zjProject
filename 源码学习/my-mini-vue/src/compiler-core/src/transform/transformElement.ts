import { NodeTypes } from "../../../parse-core/src/ast";
import { CREATE_ELEMENT_BLOCK } from "../runtimeHelpers";

export function transformElement(node, context) {
  if (node.type === NodeTypes.ELEMENT) {
    context.helper(CREATE_ELEMENT_BLOCK);
  }
}
