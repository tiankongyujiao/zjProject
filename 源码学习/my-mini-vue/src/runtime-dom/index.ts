import { createRenderer } from "../runtime-core/render";

export function createElement(type) {
  return document.createElement(type);
}

export function patchProp(el, key, oldVal, newVal) {
  const isOn = (key: string) => /^on[A-Z]/.test(key);
  if (isOn(key)) {
    const event = key.slice(2).toLowerCase();
    el.addEventListener(event, newVal);
  } else {
    if (newVal === undefined || newVal === null) {
      el.removeAttribute(key);
    } else {
      el.setAttribute(key, newVal);
    }
  }
}

export function insert(child, parent, anchor) {
  parent.insertBefore(child, anchor || null);
}

function remove(child) {
  const parent = child.parentNode;
  if (parent) {
    parent.removeChild(child);
  }
}

function setElementText(el, text) {
  el.textContent = text;
}

export let renderer: any = createRenderer({
  createElement,
  patchProp,
  insert,
  remove,
  setElementText,
});

export function createApp(...args) {
  return renderer.createApp(...args);
}

export * from "../runtime-core";
