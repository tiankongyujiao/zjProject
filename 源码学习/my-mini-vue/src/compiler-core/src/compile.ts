import { baseParse } from "../../parse-core/src/parse";
import { generate } from "./codegen";
import { transform } from "./transform";
import { transformElement } from "./transform/transformElement";
import { transformExpression } from "./transform/transformExpression";
import { transformText } from "./transform/transformText";

export function baseCompile(template) {
  const ast = baseParse("<div>hi, {{message}}</div>");
  transform(ast, {
    nodeTransforms: [transformExpression, transformElement, transformText],
  });
  return generate(ast);
}
