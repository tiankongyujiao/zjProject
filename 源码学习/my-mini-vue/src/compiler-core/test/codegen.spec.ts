import { baseParse } from "../../parse-core/src/parse";
import { generate } from "../src/codegen";
import { transform } from "../src/transform";
import { transformExpression } from "../src/transform/transformExpression";
import { transformElement } from "../src/transform/transformElement";
import { transformText } from "../src/transform/transformText";

describe("codegen", () => {
  it("string", () => {
    const ast = baseParse("hi");
    transform(ast);
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  it("interpolation", () => {
    const ast = baseParse("{{message}}");
    transform(ast, {
      nodeTransforms: [transformExpression],
    });
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });

  // it("element", () => {
  //   const ast = baseParse("<div></div>");
  //   transform(ast, {
  //     nodeTransforms: [transformElement],
  //   });
  //   const { code } = generate(ast);
  //   expect(code).toMatchSnapshot();
  // });

  it("element2", () => {
    const ast = baseParse("<div>hi, {{message}}</div>");
    transform(ast, {
      nodeTransforms: [transformExpression, transformElement, transformText],
    });
    // console.log(ast, ast.codegenNode.children, "-~~~~~~~~~~~--");
    const { code } = generate(ast);
    expect(code).toMatchSnapshot();
  });
});
