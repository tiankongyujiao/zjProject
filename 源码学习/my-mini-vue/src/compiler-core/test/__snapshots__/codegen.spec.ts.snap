// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`codegen element2 1`] = `
"const { _toDisplayString: __toDisplayString,_createElementVNode: __createElementVNode } = Vue
return function render(_ctx, _cache) {return _createElementVNode('div', null, 'hi, ' + _toDisplayString(_ctx.message))}"
`;

exports[`codegen interpolation 1`] = `
"const { _toDisplayString: __toDisplayString } = Vue
return function render(_ctx, _cache) {return _toDisplayString(_ctx.message)}"
`;

exports[`codegen string 1`] = `"return function render(_ctx, _cache) {return 'hi'}"`;
